File: src/MetaObject.ts
import { MetaObjectDraw } from "./MetaObjectDraw";
import { MetaObjectExport } from "./MetaObjectExport";
import { MetaObjectImage } from "./MetaObjectImage";
import { MetaObjectPolygon } from "./MetaObjectPolygon";
import { MetaObjectText } from "./MetaObjectText";
import { AlignmentFlags, MetaObjectType, TextAlignType, TextAlignVerticalType } from "./MetaTypes";
import { StreamUtil } from "./StreamUtil";

export abstract class MetaObject {
    public static CreateFromBuf(buf: number[], index: number): MetaObject {
        let aresult: MetaObject = null;
        const metatype: MetaObjectType = buf[index + 16] as MetaObjectType;
        switch (metatype) {
            case MetaObjectType.Text:
                aresult = new MetaObjectText();
                break;
            case MetaObjectType.Draw:
                aresult = new MetaObjectDraw();
                break;
            case MetaObjectType.Image:
                aresult = new MetaObjectImage();
                break;
            case MetaObjectType.Polygon:
                aresult = new MetaObjectPolygon();
                break;
            case MetaObjectType.Export:
                aresult = new MetaObjectExport();
                break;
        }
        return aresult;
    }
    public static GetIntHorizAlignment(Alignment: TextAlignType): number {
        let aresult: number = 0;
        if (Alignment === TextAlignType.Right)
            aresult = AlignmentFlags.AlignRight;
        else
            if (Alignment === TextAlignType.Center)
                aresult = AlignmentFlags.AlignHCenter;
            else
                if (Alignment === TextAlignType.Justify) {

                    aresult = AlignmentFlags.AlignHJustify;
                }
        return aresult;
    }
    public static GetIntVertAlignment(VAlignment: TextAlignVerticalType): number {
        // inverse the alignment for BidiMode Full
        let aresult: number = 0;
        if (VAlignment === TextAlignVerticalType.Center)
            aresult = AlignmentFlags.AlignVCenter;
        else
            if (VAlignment === TextAlignVerticalType.Bottom)
                aresult = AlignmentFlags.AlignBottom;
        return aresult;
    }
    public Id: number;
    // public const int RECORD_SIZE = 66;
    // protected static byte[] emptybuf = new byte[100];
    public Top: number;
    public Left: number;
    public Width: number;
    public Height: number;
    public MetaType: MetaObjectType;
    public FillFromBuf(buf: ArrayBuffer, index: number): void {
        this.Top = StreamUtil.byteArrayToInt(buf, index + 0);
        this.Left = StreamUtil.byteArrayToInt(buf, index + 4);
        this.Width = StreamUtil.byteArrayToInt(buf, index + 8);
        this.Height = StreamUtil.byteArrayToInt(buf, index + 12);
        this.MetaType = StreamUtil.byteArrayToByte(buf, index + 16) as MetaObjectType;
    }
    public SaveToStream(/*Stream astream*/buf: number[]): void {
        /*			astream.Write(StreamUtil.IntToByteArray(Top), 0, 4);
                    astream.Write(StreamUtil.IntToByteArray(Left), 0, 4);
                    astream.Write(StreamUtil.IntToByteArray(Width), 0, 4);
                    astream.Write(StreamUtil.IntToByteArray(Height), 0, 4);
                    astream.Write(StreamUtil.ByteToByteArray((byte)MetaType), 0, 1);*/
    }
}

File: Reportman.Drawing/MetaObject.cs
using System.IO;

namespace Reportman.Drawing
{
    /// <summary>
    /// Base class for printable objects inside a MetaPage
    /// </summary>
	public abstract class MetaObject
    {
        /// <summary>
        /// Unique identifier used by MetaFile search functions
        /// </summary>
        public long Id;
        /// <summary>
        /// Record size, all objects share the same size
        /// </summary>
		public const int RECORD_SIZE_4 = 74;
        public const int RECORD_SIZE_3 = 66;
        public const int RECORD_OFFSET_3 = 17;
        public const int RECORD_OFFSET_4 = 25;
        /// <summary>
        /// Internally used buffer
        /// </summary>
		protected static byte[] emptybuf = new byte[100];
        /// <summary>Top position of the object in twips</summary>
		public int Top;
        /// <summary>Left position of the object in twips</summary>
        public int Left;
        /// <summary>Width of the object in twips</summary>
        public int Width;
        /// <summary>height of the object in twips</summary>
        public int Height;
        /// <summary>Annotation position in MetaPage strings pool</summary>
        public int AnnotationP;
        /// <summary>Annotation length</summary>
        public int AnnotationS;
        /// <summary>Type of object</summary>
        public MetaObjectType MetaType;
        /// <summary>
        /// Creates a MetaObject, loading it from a binary buffer
        /// </summary>
        /// <param name="buf">Buffer containing information in binary format</param>
        /// <param name="index">Index to begin read in the buffer</param>
        /// <returns>Create a MetaObject using the information in the buffer</returns>
		static public MetaObject CreateFromBuf(byte[] buf, int index, int ver)
        {
            int RECORD_OFFSET = ver == 4 ? RECORD_OFFSET_4 : RECORD_OFFSET_3;
            MetaObject aresult = null;
            MetaObjectType metatype = (MetaObjectType)buf[index + RECORD_OFFSET - 1];
            switch (metatype)
            {
                case MetaObjectType.Text:
                    aresult = new MetaObjectText();
                    break;
                case MetaObjectType.Draw:
                    aresult = new MetaObjectDraw();
                    break;
                case MetaObjectType.Image:
                    aresult = new MetaObjectImage();
                    break;
                case MetaObjectType.Polygon:
                    aresult = new MetaObjectPolygon();
                    break;
                case MetaObjectType.Export:
                    aresult = new MetaObjectExport();
                    break;
                default:
                    throw new System.Exception("Tipo de meta objeto desconocido: " + metatype.ToString());
            }
            return aresult;
        }
        public static int GetRECORD_OFFSET(int ver)
        {

            int RECORD_OFFSET = ver == 4 ? RECORD_OFFSET_4 : RECORD_OFFSET_3;
            return RECORD_OFFSET;
        }
        public static int GetRECORD_SIZE(int ver)
        {

            int RECORD_SIZE = ver == 4 ? RECORD_SIZE_4 : RECORD_SIZE_3;
            return RECORD_SIZE;
        }
        public static int GetIntHorizAlignment(TextAlignType Alignment)
        {
            // Inverse the alignment for BidiMode Full
            int aresult = 0;
            if (Alignment == TextAlignType.Right)
                aresult = MetaFile.AlignmentFlags_AlignRight;
            else
                if (Alignment == TextAlignType.Center)
                aresult = MetaFile.AlignmentFlags_AlignHCenter;
            else
                    if (Alignment == TextAlignType.Justify)
                aresult = MetaFile.AlignmentFlags_AlignHJustify;
            return aresult;
        }
        public static int GetIntVertAlignment(TextAlignVerticalType VAlignment)
        {
            // Inverse the alignment for BidiMode Full
            int aresult = 0;
            if (VAlignment == TextAlignVerticalType.Center)
                aresult = MetaFile.AlignmentFlags_AlignVCenter;
            else
                if (VAlignment == TextAlignVerticalType.Bottom)
                aresult = MetaFile.AlignmentFlags_AlignBottom;
            return aresult;
        }
        /// <summary>
        /// Fill the values of a MetaObject, loading it from a binary buffer
        /// </summary>
        /// <param name="buf">Buffer containing information in binary format</param>
        /// <param name="index">Index to begin read in the buffer</param>
        virtual public void FillFromBuf(byte[] buf, int index, int ver)
        {
            int RECORD_OFFSET = GetRECORD_OFFSET(ver);
            Top = StreamUtil.ByteArrayToInt(buf, index + 0, 4);
            Left = StreamUtil.ByteArrayToInt(buf, index + 4, 4);
            Width = StreamUtil.ByteArrayToInt(buf, index + 8, 4);
            Height = StreamUtil.ByteArrayToInt(buf, index + 12, 4);
            if (ver >= 4)
            {
                AnnotationP = StreamUtil.ByteArrayToInt(buf, index + 16, 4);
                AnnotationS = StreamUtil.ByteArrayToInt(buf, index + 20, 4);
            }
            MetaType = (MetaObjectType)buf[index + RECORD_OFFSET - 1];
        }
        /// <summary>
        /// Save the information of the object into a stream
        /// </summary>
        /// <param name="astream">Destination stream</param>
		virtual public void SaveToStream(Stream astream, int ver)
        {
            astream.Write(StreamUtil.IntToByteArray(Top), 0, 4);
            astream.Write(StreamUtil.IntToByteArray(Left), 0, 4);
            astream.Write(StreamUtil.IntToByteArray(Width), 0, 4);
            astream.Write(StreamUtil.IntToByteArray(Height), 0, 4);
            if (ver >= 4)
            {
                astream.Write(StreamUtil.IntToByteArray(AnnotationP), 0, 4);
                astream.Write(StreamUtil.IntToByteArray(AnnotationS), 0, 4);
            }
            astream.Write(StreamUtil.ByteToByteArray((byte)MetaType), 0, 1);
        }
    }
}
